File: app/src/main/java/com/example/puyo_base_simulator/ui/home/Haipuyo.java
package com.example.puyo_base_simulator.ui.home;

import com.example.puyo_base_simulator.BuildConfig;

import java.io.BufferedReader;
import java.io.IOException;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class Haipuyo {
    List<String> content = new ArrayList<>();
    List<String> sortedContent = new ArrayList<>();
    // Singleton instance.
    private static final Haipuyo INSTANCE = new Haipuyo();
    private Haipuyo() {}
    public static Haipuyo getInstance() {
        return INSTANCE;
    }

    public void load(BufferedReader haipuyoBr, BufferedReader sortedBr) {
        try {
            for (int i = 0; i < 65536; i++) {
                content.add(haipuyoBr.readLine());
                sortedContent.add(sortedBr.readLine());
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    public String get(int seed) {
        if (BuildConfig.DEBUG && !(0 <= seed && seed <= 65535)) {
            throw new AssertionError("Assertion failed");
        }
        return content.get(seed);
    }
    public List<Integer> searchSeedWithPattern(String str) {
        int len = str.length();
        if (BuildConfig.DEBUG && len % 2 != 0) {
            throw new AssertionError("Assertion failed");
        }
        String sortedStr = pairwiseSort(str);
        List<Integer> ret = new ArrayList<>();
        for (int i = 0; i < 65536; i++) {
            if (sortedContent.get(i).startsWith(sortedStr)) {
                ret.add(i);
            }
        }
        return ret;
    }
    private class CharOrder {
        public int index;
        public char chara;
        public CharOrder(int index, char chara) {
            this.index = index;
            this.chara = chara;
        }
    }
    private class CharaOrderComparator implements Comparator<CharOrder> {
        @Override
        public int compare(CharOrder c1, CharOrder c2) {
            //noinspection ComparatorMethodParameterNotUsed
            return c1.index < c2.index ? -1 : 1;
        }
    }
    private String pairwiseSort(String str) {
        String newStr = "";
        for (int i = 0; i < str.length()/2; i++) {
            String substring = str.substring(i*2, i*2+2);
            char[]chars =substring.toCharArray();
            Arrays.sort(chars);
            String sorted = new String(chars);
            newStr = newStr.concat(sorted);
        }
        return newStr;
    }
}

File: app/src/main/java/com/example/puyo_base_simulator/ui/home/Haipuyo.kt
package com.example.puyo_base_simulator.ui.home

import com.example.puyo_base_simulator.BuildConfig
import java.io.BufferedReader
import java.io.IOException
import java.util.*

object Haipuyo {
    var content: MutableList<String> = ArrayList()
    var sortedContent: MutableList<String> = ArrayList()
    fun load(haipuyoBr: BufferedReader, sortedBr: BufferedReader) {
        try {
            for (i in 0..65535) {
                content.add(haipuyoBr.readLine())
                sortedContent.add(sortedBr.readLine())
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    operator fun get(seed: Int): String {
        if (BuildConfig.DEBUG && !(0 <= seed && seed <= 65535)) {
            throw AssertionError("Assertion failed")
        }
        return content[seed]
    }

    @JvmStatic
    fun searchSeedWithPattern(str: String): List<Int> {
        val len = str.length
        if (BuildConfig.DEBUG && len % 2 != 0) {
            throw AssertionError("Assertion failed")
        }
        val sortedStr = pairwiseSort(str)
        val ret: MutableList<Int> = ArrayList()
        for (i in 0..65535) {
            if (sortedContent[i].startsWith(sortedStr)) {
                ret.add(i)
            }
        }
        return ret
    }

    private class CharOrder(var index: Int, var chara: Char)
    private class CharaOrderComparator : Comparator<CharOrder> {
        override fun compare(c1: CharOrder, c2: CharOrder): Int {
            return if (c1.index < c2.index) -1 else 1
        }
    }

    private fun pairwiseSort(str: String): String {
        var newStr = ""
        for (i in 0 until str.length / 2) {
            val substring = str.substring(i * 2, i * 2 + 2)
            val chars = substring.toCharArray()
            Arrays.sort(chars)
            val sorted = String(chars)
            newStr = newStr + sorted
        }
        return newStr
    }
}